cpp/src/io/json/read_json.cu

```diff
@@ -68,33 +74,46 @@ rmm::device_uvector<char> ingest_raw_input(host_span<std::unique_ptr<datasource>
   auto constexpr num_delimiter_chars = 1;
   auto const num_extra_delimiters    = num_delimiter_chars * (sources.size() - 1);
 
-  // Iterate through the user defined sources and read the contents into the local buffer
-  auto const total_source_size =
-    sources_size(sources, range_offset, range_size) + num_extra_delimiters;
-
   if (compression == compression_type::NONE) {
     std::vector<size_type> delimiter_map{};
+    std::vector<size_t> prefsum_source_sizes(sources.size());
+    std::vector<std::unique_ptr<datasource::buffer>> h_buffers;
     delimiter_map.reserve(sources.size());
-    auto d_buffer     = rmm::device_uvector<char>(total_source_size, stream);
     size_t bytes_read = 0;
-    std::vector<std::unique_ptr<datasource::buffer>> h_buffers;
-    for (auto const& source : sources) {
-      if (!source->is_empty()) {
-        auto data_size   = (range_size != 0) ? range_size : source->size();
-        auto destination = reinterpret_cast<uint8_t*>(d_buffer.data()) + bytes_read;
-        if (source->is_device_read_preferred(data_size)) {
-          bytes_read += source->device_read(range_offset, data_size, destination, stream);
-        } else {
-          h_buffers.emplace_back(source->host_read(range_offset, data_size));
-          auto const& h_buffer = h_buffers.back();
-          CUDF_CUDA_TRY(cudaMemcpyAsync(
-            destination, h_buffer->data(), h_buffer->size(), cudaMemcpyDefault, stream.value()));
-          bytes_read += h_buffer->size();
-        }
-        delimiter_map.push_back(bytes_read);
-        bytes_read += num_delimiter_chars;
+    std::transform(sources.begin(),
+                   sources.end(),
+                   prefsum_source_sizes.begin(),
+                   [](const std::unique_ptr<datasource>& s) { return s->size(); });
+    std::inclusive_scan(
+      prefsum_source_sizes.begin(), prefsum_source_sizes.end(), prefsum_source_sizes.begin());
+    auto upper =
+      std::upper_bound(prefsum_source_sizes.begin(), prefsum_source_sizes.end(), range_offset);
+    size_t start_source = std::distance(prefsum_source_sizes.begin(), upper);
+
+    range_size = !range_size || range_size > prefsum_source_sizes.back()
+                   ? prefsum_source_sizes.back() - range_offset
+                   : range_size;
```

Comment:
Good catch, we don't have to actually test if range size is non-zero here because we are adjusting it in `get_record_range_raw_input` if the whole source needs to be read. I've added this fix, and also included a tests that should read till the end of the source list.